sematic/abstract_future.py

```diff
@@ -81,9 +83,56 @@ class FutureProperties:
     inline: bool
     name: str
     tags: List[str]
+    kwargs: Dict[str, Any]
     resource_requirements: Optional[ResourceRequirements] = None
     retry_settings: Optional[RetrySettings] = None
     base_image_tag: Optional[str] = None
+    external_resources: List[ExternalResource] = field(default_factory=list)
+    # We don't want to replace futures in kwargs, because it holds
+    # the source of truth for the future graph. Instead we have concrete
+    # values in resolved_kwargs
+    # It will be set only once all input values are resolved
+    # resolved_kwargs: Dict[str, Any] = field(default_factory=dict)
+
+    def get_resolved_kwargs(self) -> Dict[str, Any]:
```

Comment:
moved this from `future` down to `future.props` in keeping with the general goal of doing that with everything in `future`. Also changed to a method so there's less internal state to reason about and also we can get the latest remote resource info whenever we access this.

Reply:
changed to a method -> 👍 
moved to props -> unsure about it but not a deal breaker. It's not a prop, and it's unlikely anyone will ever need to access a `get_resolved_kwargs` attribute on a future object, it's an unlikely dataclass field name 😬

Reply:
The whole point of `Props` was just to de-clutter the namespace of `Future` as much as possible so that we could support things like this with minimal chance for naming conflicts:

```
@sematic.func
def pipeline() -> int:
    future_dataclass = do_something()
    return future_dataclass.x
```

Though perhaps `Props` is too restrictive of a name for this, since we want to have methods as well as data there. I vote we rename `Props` to `FutureAttributes` (attributes -> name covering methods and fields). Then on the future we have literally only one field, `sematic_attributes`. So you would do:
```
my_future.sematic_attributes.kwargs
```
to access kwargs, and
```
my_future.foo
```
as a shorthand for creating a func that returns the `foo` property from the resolved value in `my_future`.